### PR TITLE
Have inheritdoc for API classes

### PR DESCRIPTION
**Pull request by gerner-spryker**

### CODE REVIEW

File: src/Common/ImplementsApiInheritdocRule.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+namespace ArchitectureSniffer\Common;
+
+use ArchitectureSniffer\SprykerAbstractRule;
+use PHPMD\AbstractNode;
+use PHPMD\Node\MethodNode;
+use PHPMD\Rule\ClassAware;
+
+class ImplementsApiInheritdocRule extends SprykerAbstractRule implements ClassAware
+{
+    public const RULE = 'Every API method must also contain the @inheritdoc tag in docblock and a contract text above.';
```
Comment (dereuromark) [inline src/Common/ImplementsApiInheritdocRule.php]: inheritDoc

File: src/Common/ImplementsApiInheritdocRule.php [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ */
+
+namespace ArchitectureSniffer\Common;
+
+use ArchitectureSniffer\SprykerAbstractRule;
+use PHPMD\AbstractNode;
+use PHPMD\Node\MethodNode;
+use PHPMD\Rule\ClassAware;
+
+class ImplementsApiInheritdocRule extends SprykerAbstractRule implements ClassAware
```
Comment (dereuromark) [inline src/Common/ImplementsApiInheritdocRule.php]: InheritDoc

File: src/Common/ImplementsApiInheritdocRule.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+     * @return void
+     */
+    protected function applyEveryInterfaceMethodMustHaveApiTagAndContractText(MethodNode $method)
+    {
+        $comment = $method->getComment();
+        if (preg_match(
+            '(
+                \*\s+[{}A-Z0-9\-]+.*\s+
+                \*?\s*
+                \*\s+@inheritdoc
```
Comment (dereuromark) [inline src/Common/ImplementsApiInheritdocRule.php]: inheritDoc

File: src/Common/ruleset.xml [hunk 4]
```diff
@@ -18,6 +18,13 @@
         <priority>1</priority>
     </rule>
 
+    <rule
+            name="ImplementsApiInheritdocRule"
```
Comment (dereuromark) [inline src/Common/ruleset.xml]: inheritDoc

**Comment by gerner-spryker** [comment]
@dereuromark Please wait for the "request-review" label ;)

File: src/Common/ImplementsApiInheritDocRule.php [hunk 5]
```diff
@@ -51,35 +48,20 @@ public function apply(AbstractNode $node)
...
+    protected function applyOnMethod(MethodNode $method)
     {
         $comment = $method->getComment();
         if (preg_match(
-            '(
-                \*\s+[{}A-Z0-9\-]+.*\s+
-                \*?\s*
-                \*\s+@inheritdoc
-            )xi',
+            '/\{\@inheritDoc\}/',
```
Comment (dereuromark) [inline src/Common/ImplementsApiInheritDocRule.php]: Should be /i as we want to find it regarless of casing.

File: src/Common/ruleset.xml [hunk 6]
```diff
@@ -18,6 +18,13 @@
         <priority>1</priority>
     </rule>
 
+    <rule
+            name="ImplementsApiInheritDocRule"
+            message="Access: {0}"
+            class="ArchitectureSniffer\Common\ImplementsApiInheritDocRule">
+
+        <priority>3</priority>
```
Comment (gerner-spryker) [inline src/Common/ruleset.xml]: @dereuromark I've 1 issue here. Even with prio 3, it still causes errors. What to do here? Should I fix entire Spryker/spryker? :S

Comment (gerner-spryker) [inline src/Common/ruleset.xml]: @dereuromark Could you please help selecting the priority?

**Comment by dereuromark** [comment]
OK but please rebase this on top of master, this is a bit outdated.

File: src/Common/ruleset.xml (hunk 6 above)
Comment (dereuromark) [reply src/Common/ruleset.xml]: 3 is fine, we need to then check if we can ignore/whitelist certain modules.

**Comment by gerner-spryker** [comment]
master merged

File: src/Common/ImplementsApiInheritDocRule.php [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        if (strpos($classNamespace, 'Dependency\\') !== false) {
+            return;
+        }
+
+        $className = $classNode->getName();
+        if (!preg_match('/(' . implode('|', $this->apiClasses) . ')$/', $className)) {
+            return;
+        }
+
+        /** @var \PHPMD\Node\InterfaceNode $classNode */
```
Comment (dereuromark) [inline src/Common/ImplementsApiInheritDocRule.php]: Is this valid? I think you want multiline docblock as single one here.

File: src/Common/ImplementsApiInheritDocRule.php [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+    /**
+     * @param \PHPMD\Node\MethodNode|\PDepend\Source\AST\ASTMethod $methodNode
+     *
+     * @return void
+     */
+    protected function applyOnPublicMethod(MethodNode $methodNode)
+    {
+        $methodDocBlock = $methodNode->getComment();
+        if (preg_match('/\{\@inheritDoc\}/', $methodDocBlock)) {
```
Comment (dereuromark) [inline src/Common/ImplementsApiInheritDocRule.php]: The /i still missing here.

**Comment by dereuromark** [comment]
ping @gerner-spryker

**Comment by gerner-spryker** [comment]
We had additional issue, it couldn't be straight-forwardly released yet. That's why i didn't go forward with the issue. I will ping you as soon as I finalised everything.

File: src/Common/ImplementsApiInheritDocRule.php (hunk 7 above)
Comment (gerner-spryker) [reply src/Common/ImplementsApiInheritDocRule.php]: no, i don't

Comment (dereuromark) [reply src/Common/ImplementsApiInheritDocRule.php]: why?

File: src/Common/ImplementsApiInheritDocRule.php (hunk 8 above)
Comment (gerner-spryker) [reply src/Common/ImplementsApiInheritDocRule.php]: done

File: src/Common/ImplementsApiInheritDocRule.php (hunk 7 above)
Comment (gerner-spryker) [reply src/Common/ImplementsApiInheritDocRule.php]: done

**Pull request merged**